Replace hammerJS with native touch events

vpetrusevici opened the pull request:
Currently I implemented swiping only for touch events (not mouse) and only for gallery-swiper component (not thumbs).
[MurhafSousli](https://github.com/MurhafSousli) Can you check please if my implementation is good enough and I can continue work in this way?
For #449

MurhafSousli commented on projects/ng-gallery/src/lib/components/gallery-slider.component.ts:
```diff
@@ -172,41 +148,23 @@ export class GallerySliderComponent implements OnInit, OnChanges, OnDestroy {
-      }
+  private onSwipe(e: SwipeEvent) {
+    if (e.direction === this.config.slidingDirection) {
+      const velocity = e.distance / (e.moveEvent.sourceEvent.timeStamp - e.startEvent.sourceEvent.timeStamp);
```
Would be better if we get the velocity with the `SwipeEvent`

vpetrusevici replied:
I agree. It was here just to test :)

vpetrusevici replied:
changed

MurhafSousli reviewed (changes requested):
My review is based on the code changes, but swiping didn't work!

vpetrusevici commented:
@MurhafSousli strange about swiping. I just retested it in chrome. Looks like is working https://amdaris-my.sharepoint.com/:v:/p/vladimir_petrusevici/EShrbtcG6pFDng1Z2TeSJQ8BzEmBGQIiCwkzAaM7tEAXzg?e=UBLKuq

vpetrusevici commented on projects/ng-gallery/src/lib/components/gallery-slider.component.ts:
```diff
@@ -94,42 +93,21 @@ export class GallerySliderComponent implements OnInit, OnChanges, OnDestroy {
+        domElement: this._el.nativeElement,
+        onSwipeMove: event => {
+          if (event.direction === this.config.slidingDirection) {
+            this._zone.run(() => this.updateSlider({ value: event.distance, active: true }));
```
Ok. I'll try to avoid it

vpetrusevici replied:
Looks like hammerJS somehow triggers ChangeDetection on each pan event, but here I don't see any possibility to trigger async pipe outside of ngZone or without manually call detection. Here an example https://stackoverflow.com/a/35513390 
Mb we can avoid ngStyles and use native js styles change, but don't think it's a good solution

vpetrusevici replied:
I'll continue with mouse events and thumbs swiping for now. Let me know if you have any ideas how to avoid change detection here.

vpetrusevici replied:
As i see we can improve performance using custom async pipe with subscription outside of ngZone as discussed in https://github.com/angular/angular/issues/36139 
Example: https://stackblitz.com/edit/angular-async-pipe-outside-angular?file=src%2Fapp%2Fvo-async.pipe.ts

vpetrusevici replied:
I replaced all async pipes with this custom pipe, and yes, it's really effective. Now we can create an observable outside of ngZone and don't care about changeDetection.

vpetrusevici commented on projects/ng-gallery/src/lib/components/gallery-slider.component.ts:
```diff
@@ -94,42 +93,21 @@ export class GallerySliderComponent implements OnInit, OnChanges, OnDestroy {
   }
 
   ngOnInit() {
-    if (this.config.gestures && typeof Hammer !== 'undefined') {
```
cahnged

vpetrusevici commented:
Also I see a problem with autoplay. Application will never have a stable state when autoPlay = true in the current implementation.
You can check it in this way 
![image](https://user-images.githubusercontent.com/40790510/183844343-816920b8-8259-43a3-a453-30c3215174de.png)

I think this observable with delay should be also moved outside ngZone.

vpetrusevici commented:
> Also I see a problem with autoplay. Application will never have a stable state when autoPlay = true in the current implementation. You can check it in this way ![image](https://user-images.githubusercontent.com/40790510/183844343-816920b8-8259-43a3-a453-30c3215174de.png)
> 
> I think this observable with delay should be also moved outside ngZone.

Fixed with custom async pipe. Now the app is always stable.

vpetrusevici commented:
Implemented mouse events

vpetrusevici commented:
Aaaand looks like all functionalities are implemented now.

vpetrusevici commented:
@MurhafSousli  should i change the docs or is it better to leave it to you?

vpetrusevici commented:
I should stop pushing... sorry)

vpetrusevici commented:
@MurhafSousli Hi, I fixed the conflicts. Can you check this PR please?

vpetrusevici commented:
Hi @MurhafSousli,
 did you have a chance to check it out? Сan't wait to finish this PR :)

MurhafSousli commented:
> Hi @MurhafSousli, did you have a chance to check it out? Сan't wait to finish this PR :)

When you add many unnecessary changes, you make it difficult for me to review! please keep the PR focused next time and don't refactor parts that are not related to the replacement of hammer.js

vpetrusevici commented:
> > Hi @MurhafSousli, did you have a chance to check it out? Сan't wait to finish this PR :)
> 
> When you add many unnecessary changes, you make it difficult for me to review! please keep the PR focused next time and don't refactor parts that are not related to the replacement of hammer.js

Sorry, i can try to recreate the PR with only necessary changes, but don't think it will be much smaller

MurhafSousli commented:
I think HammerJS is not bad, comparing the benefits we would get from implementing our own touch-events solution does not bring any value in terms of feature or functionality, we are just adding more code which we will have to maintenance, in favor of getting rid of one dependency.

I will close this for now, but will keep an eye on your solution if something changed.

Thank you for PR anyway!
